File: src/record.ts
import { basename } from "node:path";
import { dumpEncIdFromName, isEncryptedDumpName } from "./dumps.ts";
import type { Session } from "./metadata.ts";
import { encId } from "./metadata.ts";
import {
  Action,
  getDumpByPath,
  recordAudit,
  recordRestore,
  registerDump,
  relativeDumpPath,
  sha256File,
  Status,
} from "./vault.ts";

export async function recordDump(
  session: Session | null,
  dumpsRoot: string,
  absPath: string,
  itemKey: string,
): Promise<void> {
  if (!session) return;
  const rel = relativeDumpPath(dumpsRoot, absPath);
  const { hash, size } = await sha256File(absPath);
  const name = basename(absPath);
  const enc = isEncryptedDumpName(name);
  let encIdVal = dumpEncIdFromName(name) ?? "";
  if (!encIdVal) encIdVal = encId(session);
  registerDump(session.db, rel, itemKey, name, hash, size, enc, encIdVal);
  recordAudit(
    session.db,
    Action.Dump,
    Status.Success,
    itemKey,
    rel,
    `sha256=${hash} size=${size}`,
    "",
  );
}

export async function recordRestoreOp(
  session: Session | null,
  dumpsRoot: string,
  absPath: string,
  destKey: string,
  durationMs: number,
  clean: boolean,
  warnings: string,
  restoreErr: Error | null,
): Promise<void> {
  if (!session) return;
  let rel = "";
  try {
    rel = relativeDumpPath(dumpsRoot, absPath);
  } catch {
    rel = basename(absPath);
  }
  let hash = "";
  try {
    const h = await sha256File(absPath);
    hash = h.hash;
  } catch {
    // ignore missing file
  }
  const rec = getDumpByPath(session.db, rel);
  recordRestore(session.db, {
    dumpId: rec?.id ?? null,
    dumpRelativePath: rel,
    dumpSha256: hash,
    destinationKey: destKey,
    durationMs,
    status: restoreErr ? Status.Failure : Status.Success,
    cleanRestore: clean,
    warnings,
    errorMessage: restoreErr?.message ?? "",
  });
  if (restoreErr) {
    recordAudit(session.db, Action.Restore, Status.Failure, rel, destKey, "", restoreErr.message);
  } else {
    recordAudit(
      session.db,
      Action.Restore,
      Status.Success,
      rel,
      destKey,
      `duration_ms=${durationMs} clean=${clean}`,
      "",
    );
  }
}

export function recordSyncAudit(
  session: Session | null,
  sourceKey: string,
  destKey: string,
  dumpRel: string,
  err: Error | null,
): void {
  if (!session) return;
  recordAudit(
    session.db,
    Action.Sync,
    err ? Status.Failure : Status.Success,
    sourceKey,
    destKey,
    dumpRel,
    err?.message ?? "",
  );
}

File: src/internal/app/record.go
package app

import (
	"fmt"
	"path/filepath"

	"github.com/md-redwan-hossain/dumpmgr/src/internal/dumps"
	"github.com/md-redwan-hossain/dumpmgr/src/internal/metadata"
	"github.com/md-redwan-hossain/dumpmgr/src/internal/vault"
)

func recordDump(session *metadata.Session, dumpsRoot, absPath, itemKey string) error {
	if session == nil || session.Store == nil {
		return nil
	}
	rel, err := vault.RelativeDumpPath(dumpsRoot, absPath)
	if err != nil {
		return err
	}
	hash, size, err := vault.SHA256File(absPath)
	if err != nil {
		return err
	}
	name := filepath.Base(absPath)
	enc := dumps.IsEncryptedDumpName(name)
	encID := dumps.DumpEncIDFromName(name)
	if encID == "" {
		if id, err := metadata.EncID(session); err == nil {
			encID = id
		}
	}
	if _, err := session.Store.RegisterDump(rel, itemKey, name, hash, size, enc, encID); err != nil {
		return err
	}
	return session.Store.RecordAudit(vault.ActionDump, vault.StatusSuccess, itemKey, rel, fmt.Sprintf("sha256=%s size=%d", hash, size), "")
}

func recordRestore(session *metadata.Session, dumpsRoot, absPath, destKey string, durationMS int64, clean bool, warnings string, restoreErr error) {
	if session == nil || session.Store == nil {
		return
	}
	rel, _ := vault.RelativeDumpPath(dumpsRoot, absPath)
	hash, _, _ := vault.SHA256File(absPath)
	var dumpID *int64
	if rec, _ := session.Store.GetDumpByPath(rel); rec != nil {
		id := rec.ID
		dumpID = &id
	}
	status := vault.StatusSuccess
	errMsg := ""
	if restoreErr != nil {
		status = vault.StatusFailure
		errMsg = restoreErr.Error()
	}
	cleanInt := clean
	_ = session.Store.RecordRestore(vault.RestoreRecord{
		DumpID:           dumpID,
		DumpRelativePath: rel,
		DumpSHA256:       hash,
		DestinationKey:   destKey,
		DurationMS:       durationMS,
		Status:           status,
		CleanRestore:     cleanInt,
		Warnings:         warnings,
		ErrorMessage:     errMsg,
	})
	action := vault.ActionRestore
	subject := rel
	if status == vault.StatusSuccess {
		_ = session.Store.RecordAudit(action, status, subject, destKey, fmt.Sprintf("duration_ms=%d clean=%v", durationMS, clean), "")
	} else {
		_ = session.Store.RecordAudit(action, status, subject, destKey, "", errMsg)
	}
}

func recordSyncAudit(session *metadata.Session, sourceKey, destKey, dumpRel string, err error) {
	if session == nil || session.Store == nil {
		return
	}
	status := vault.StatusSuccess
	errMsg := ""
	if err != nil {
		status = vault.StatusFailure
		errMsg = err.Error()
	}
	_ = session.Store.RecordAudit(vault.ActionSync, status, sourceKey, destKey, dumpRel, errMsg)
}
